docs/timezone-draft.md

```diff
@@ -0,0 +1,169 @@
+
+When parsing an ISO 8601 string, the only places the time zone identifier is taken into account are `Temporal.Absolute.from()` and `Temporal.TimeZone.from()`.
+
+These functions gain an `options` parameter with an `idToTimeZone` option, whose value is a function returning the appropriate time zone object for the `id` passed in, or `null` to indicate that the time zone doesn't exist.
+
+If the option is not given, `Temporal.TimeZone.idToTimeZone()` is called.
+This is a new static method of `Temporal.TimeZone`.
+
+> **FIXME:** Note that there is an issue open regarding this behaviour for calendars: [#294](https://github.com/tc39/proposal-temporal/issues/294)
+
+```javascript
+function idToTimeZone(id) {
+  if (id === 'Mars/Olympus_Mons')
+    return new MarsTime();
+  return Temporal.TimeZone.idToTimeZone(id);
+}
+
+Temporal.Absolute.from('2020-04-01T18:16+00:00[Mars/Olympus_Mons]', { idToTimeZone })
+  // returns the Absolute corresponding to 2020-04-01T18:16Z,
+  // assuming that MarsTime instances have an offset of +00:00 at that instant
+
+Temporal.TimeZone.from('2020-04-01T18:16+00:00[Mars/Olympus_Mons]', { idToTimeZone })
+  // returns a new MarsTime instance
+```
+
+> **FIXME:** Passing a custom time zone's identifier to the built-in `Temporal.TimeZone` constructor currently doesn't work: `new Temporal.TimeZone('Mars/Olympus_Mons')` throws.
+> However, this must change, because implementations would need to call `super(id)` to set the _[[Identifier]]_ and _[[InitializedTemporalTimeZone]]_ internal slots.
+> Maybe we need to make `Temporal.TimeZone` not directly constructable?
+
+In order to perform attenuation, an implementation would monkeypatch the `Temporal.TimeZone.idToTimeZone()` function which performs the built-in mapping, and change the list of allowed time zones that `Temporal.TimeZone` iterates through.
+Or it could just replace the `Temporal.TimeZone` class altogether:
+
+```javascript
+// For example, to allow only offset time zones:
+
+TemporalTimeZone_idToTimeZone = Temporal.TimeZone.idToTimeZone;
+Temporal.TimeZone.idToTimeZone = function (id) {
+  if (/^[+-]\d{2}:?\d{2}$/.test(id))
+    return TemporalTimeZone_idToTimeZone(id);
+  return null;
+}
+Temporal.TimeZone[Symbol.iterator] = function* () { return null; }
+
+// or, to replace the built-in implementation altogether:
+
+Temporal.TimeZone = FrozenTimeZoneImplementation;
+```
+
+## Implementation of a custom time zone
+
+### TimeZone interface
+
+```javascript
+class Temporal.TimeZone {
+  /** Sets the [[Identifier]] internal slot to @id, and creates the
+   * [[InitializedTemporalTimeZone]] internal slot */
+  constructor(id : string) : Temporal.TimeZone;
+
+  // Methods that a subclassed custom time zone must implement
+
+  /** Given an instant (in epoch nanoseconds) returns this time zone's
+   * corresponding UTC offset, in nanoseconds (signed). */
+  getOffsetAtInstant(epochNs : bigint) : bigint;
```

Comment:
Why not pass a Temporal.Absolute here?

Reply:
I could go either way on this. I guess if we make it public then it should take an Absolute.

One minor consideration for not passing Temporal objects was to stop implementors from creating infinite loops by calling Absolute.inTimeZone or DateTime.inTimeZone in these methods, but I agree that if we consider the methods public then the other reasons for passing Temporal objects outweigh this consideration. Anyway, someone who made that mistake would find out soon enough.

Reply:
The infinite loop was the same issue I was raising as a potential concern in #354.  We decided there that it was safe and sufficient to have methods like `.getISOFields()` that don't call into the calendar.  I think the same line of thinking should apply here.

This is only relevant when implementing custom time zones, so it's a bit of an edge case.